servicemanager/servicemanager.go

```diff
@@ -259,39 +358,146 @@ func (sm *ServiceManager) doApplyService(service ServiceInfo) (err error) {
 		return aoserrors.Wrap(err)
 	}
 
+	sm.Lock()
+
 	for _, oldService := range oldServices {
 		if oldService.AosVersion == service.AosVersion {
 			continue
 		}
 
-		if err := os.RemoveAll(oldService.ImagePath); err != nil {
+		freedServiceSize, _, err := sm.removeService(oldService)
+		if err != nil {
 			log.Errorf("Can't remove old service: %s", err)
+			continue
 		}
 
-		if err := sm.serviceInfoProvider.RemoveService(oldService); err != nil {
-			log.Errorf("Can't remove old service from storage: %s", err)
+		sm.availableSize += freedServiceSize
+	}
+
+	sm.Unlock()
+
+	if err = sm.serviceInfoProvider.ActivateService(service.ServiceID, service.AosVersion); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (sm *ServiceManager) doRevertService(service ServiceInfo) error {
+	sm.Lock()
+	defer sm.Unlock()
+
+	freedServiceSize, _, err := sm.removeService(service)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	sm.availableSize += freedServiceSize
+
+	return nil
+}
+
+func (sm *ServiceManager) removeOutdateServices(extraSpace int64) (deallocatedSize int64, err error) {
+	sm.Lock()
+	defer sm.Unlock()
+
+	for deallocatedSize < extraSpace {
+		if len(sm.pendingRemove) == 0 {
+			return 0, aoserrors.New("not enough memory")
 		}
+
+		freedServiceSize, freedLayersSize, err := sm.removeService(sm.pendingRemove[0])
+		if err != nil {
+			return 0, aoserrors.Wrap(err)
+		}
+
+		sm.availableSize += freedServiceSize
+
+		deallocatedSize += freedLayersSize
+
+		sm.pendingRemove = sm.pendingRemove[1:]
 	}
 
-	if err = sm.serviceInfoProvider.ActivateService(service); err != nil {
+	return deallocatedSize, nil
+}
+
+func (sm *ServiceManager) doRemoveService(serviceID string) error {
+	service, err := sm.serviceInfoProvider.GetService(serviceID)
+	if err != nil {
 		return aoserrors.Wrap(err)
 	}
 
+	sm.Lock()
+	defer sm.Unlock()
+
+	if service.Timestamp.Add(time.Hour * 24 * time.Duration(sm.serviceTTLDays)).Before(time.Now()) {
+		freedServiceSize, _, err := sm.removeService(service)
+		if err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		sm.availableSize += freedServiceSize
+	} else {
+		sm.pendingRemove = append(sm.pendingRemove, service)
```

Comment:
pendingRemove should be sorted by timestamp, most recently used services should be deleted in last row. So, consider using list for it.

Reply:
Done